Core/Core/UITestHelpers/IPC.swift

```diff
@@ -0,0 +1,114 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2019-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+#if DEBUG
+
+import Foundation
+
+protocol IPCServer {
+    var machPortName: String { get }
+    var messagePort: CFMessagePort { get }
+    var queue: DispatchQueue { get }
+
+    func handler(msgid: Int32, data: Data?) -> Data?
+}
+
+private var knownIPCServers = [CFMessagePort: IPCServer]()
+extension IPCServer {
+    func makePort() -> CFMessagePort {
+        let handlerWrapper: CFMessagePortCallBack = { port, msgid, data, _ in
+            knownIPCServers[port!]!.handler(msgid: msgid, data: data as Data?).map { Unmanaged.passRetained($0 as CFData) }
+        }
+        guard let port = CFMessagePortCreateLocal(kCFAllocatorDefault, machPortName as CFString, handlerWrapper, nil, nil) else {
+            fatalError("Couldn't create mach port \(machPortName)")
+        }
+        knownIPCServers[port] = self
+        CFMessagePortSetDispatchQueue(port, queue)
+        return port
+    }
+}
+
+//class IPCTestDriverServer: IPCServer {
+//    let machPortName = "com.instructure.icanvas.ui-test-driver"
+//    lazy var messagePort: CFMessagePort = makePort()
+//    lazy var queue: DispatchQueue = DispatchQueue(label: machPortName)
+//
+//    func handler(msgid: Int32, data: Data?) -> Data? { fatalError("handler(msgid:data:) has not been implemented") }
+//}
+//
```

Comment:
Are we not quite ready to push things back up to the test driver?

Reply:
It's ready, just not needed yet. I wasn't sure whether to add it in unused or leave it out.